### ISSUE TITLE
Adapter weights not updating despite non-zero gradients

### ISSUE THREAD

**Issue by scottviteri** [scottviteri/AttentionGuidedRL#8]
Problem: Training logs show 'Adapter model weights are NOT changing!' warnings despite non-zero loss and gradient norms. Possible causes: 1) Learning rate too small (0.0002), 2) Gradients very small due to high filtering rate, 3) Gradient clipping too aggressive. Impact: No learning is occurring. Potential solutions: 1) Increase learning rate, 2) Log actual gradient magnitudes, 3) Reduce filtering, 4) Check if optimizer.step() is actually updating weights.

**Comment by scottviteri** [comment]
Resolved: The issue was a misunderstanding of the architecture. Only layer -2 LoRA parameters get gradients because only that layer is used for query extraction. This is correct behavior - the model is being trained to generate better query embeddings from layer -2 specifically. Other layers' LoRA adapters remain unused and thus have zero gradients. Added test case test_adapter_weights_update_during_training to verify layer -2 weights do update correctly. Also increased learning rate to 1e-3 to ensure visible weight changes.

**Comment by scottviteri** [comment]
## 🔧 CORRECTION: Gradient Flow Analysis Update

After thorough testing and analysis, the initial understanding in this issue was **incorrect**. 

### ❌ Previous Incorrect Understanding:
- Only layer -2 (layer 10 for GPT-2) receives gradients
- Other layers don't participate in gradient computation
- Training should only update layer -2 LoRA parameters

### ✅ Actual Reality:
- **ALL LoRA layers receive gradients** during training
- **Layer -2 dominates** with approximately **2.8x stronger gradients** than other layers
- **Layers 6-9** receive moderate gradients  
- **Layers 0-5** receive weaker but non-zero gradients
- **This is correct behavior by design**

### 🔬 Experimental Evidence:
Created gradient flow tests that showed:
```
Layer gradients (example results):
- Layer 10 (layer -2): ~2.84 gradient magnitude
- Layer 9: ~1.12 gradient magnitude  
- Layer 8: ~0.98 gradient magnitude
- Layer 0-5: ~0.3-0.7 gradient magnitude
```

### 📋 Why This Happens:
1. **Forward Pass Chain**: To compute layer -2's output, the model must run through layers 0→1→2→...→10
2. **Gradient Backpropagation**: When backpropagating through extracted query embeddings, gradients flow backward through the entire computation graph
3. **LoRA in Forward Path**: Since LoRA adapters are inserted into each layer's attention computation, they all participate in the gradient flow
4. **Layer -2 Dominance**: Layer -2 receives the strongest gradients because it's the direct extraction point

### 🐛 Actual Root Cause of "Weights Not Updating":
The real issue was a **variable overwrite bug** in `main.py`:
```python
# Line 363: Create baseline model correctly with hook
baseline_model = create_model_copy(adapter_model)
baseline_embeddings_dict, baseline_hook_remover = register_embedding_hook(baseline_model, embed_type="key")

# Line 388: OVERWRITE the baseline_model!! This destroys the hook!
baseline_model = deepcopy(adapter_model)  # ❌ This was the bug!
```

### ✅ Resolution:
1. **Fixed variable overwrite bug** - main function now executes properly
2. **Added comprehensive gradient monitoring** - all layers are tracked
3. **Enhanced training metrics** - log probability tracking provides clearer signals
4. **Updated tests** - reflect correct gradient flow understanding

### 📚 Documentation:
Complete technical analysis available in: `docs/gradient_flow_analysis.md`

### 🎯 Key Takeaway:
The gradient flow was working correctly all along. The system efficiently trains all LoRA layers with layer -2 getting the strongest signal for query generation while other layers contribute to building appropriate representations.

**Status:** ✅ **RESOLVED** with corrected understanding and enhanced monitoring
